Add runtime type checking in CI tests

DimitriPapadopoulos opened the pull request:
This should prevent `TYPE_CHECKING` misuse.

Fixes #9744.

DimitriPapadopoulos committed:
Add runtime type checking in CI tests

This should prevent TYPE_CHECKING misuse.

It might also found issues not found by static type checking,
typically by the current MyPy.

Make it optional for now, while we fix issues.

keewis commented on .github/workflows/ci.yaml:
```diff
@@ -152,6 +156,7 @@ jobs:
           --cov=xarray
           --cov-report=xml
           --junitxml=pytest.xml
+          --beartype-packages=xarray
```
can we move the runtime type-checking to a separate job in `ci-additional`? I don't think we *have* to perform runtime type checking as extensively as the main CI, plus I'd like to have test failures separate from type-checking errors.

DimitriPapadopoulos replied:
Note that beartype is supposed to be lightwheight:
> Unbearably fast near-real-time hybrid runtime-static type-checking in pure Python. 

Anyway, I too wouldn't like to mix type checking with other tests for now. The best fit is probably [.github/workflows/ci-additional.yaml](https://github.com/pydata/xarray/blob/main/.github/workflows/ci-additional.yaml) together with PyPy static type checking.

max-sixty commented:
This is a good test! Unfortunately our CI fails, and we couldn't work out a succinct way of changing our annotations so that we could resolve #9581.

If there were some way of adding this in part, that would be great. But likely we'll need to wait for more support so we can support this while having just `DataArray` in the type definition without the full path (more discussion in the attached issue)

headtr1ck commented:
Maybe we make it optional, similar to pyright?

DimitriPapadopoulos commented:
I was planning on exploring the suggestions in https://github.com/pydata/xarray/issues/9581.

But yes, I could make the test optional for now.

DimitriPapadopoulos committed:
TESTING

max-sixty commented:
Is it helpful to have an optional CI which fails everytime at the first import though?

DimitriPapadopoulos commented:
It looks like solutions that might work with beartype fail with MyPy. I haven't been able to find a solution within the little time I can dedicate to this issue.

In any case, it looks like `TYPE_CHECKING` shouldn't be used, except to avoid circular imports.

DimitriPapadopoulos committed:
Avoid PEP 585 and PEP 604 rewrites

This setting is only applicable when the target Python version is
below 3.9 and 3.10 respectively.
